File: Custom Sort String.py
# Time Complexity: O(n)
# Space Complexity: O(1)
class Solution:
    def customSortString(self, order: str, s: str) -> str:
        hmap = collections.defaultdict(int)
        res = ''
        for ch in s:
            hmap[ch] += 1

            # print(hmap)

        for i in order:
            res += i * hmap[i]
            hmap[i] = 0

        for key, val in hmap.items():
            res += key * val

        return res

File: custom-sort-string.cpp
// Time Complexity : O(nlogn)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes

#include <iostream>
#include <vector>
#include <algorithm> // for sort

std::string customSortString_usingCustomSort(std::string order, std::string str) {
    int sortedOrder[26] = {0};

    for (int i = 0; i < order.length(); ++i) {
        sortedOrder[order[i] - 'a'] = i;
    }

    std::vector<char> arr(str.length());
    for (int i = 0; i < str.length(); ++i) {
        arr[i] = str[i];
    }

    std::sort(arr.begin(), arr.end(), [&](char c1, char c2) {
        return sortedOrder[c1 - 'a'] < sortedOrder[c2 - 'a'];
    });

    std::string result;
    for (char c : arr) {
        result.push_back(c);
    }

    return result;
}